### PR TITLE
feat(mpo): Add MPO agent implementation

### PR DESCRIPTION
**Pull request by ethanluoyc**
Add a JAX MPO agent implementation based on acme/agents/tf/mpo.

Some notes on the implementation

1. It uses a single pass for computing the gradients wrt
   policy and critic using vjp. This is similar to the way
   a single pass is used for computing gradients in the jax
   value_dice agent and is the closest to what is being done
   in the tf mpo agent.
2. It uses a single network for the policy network and does not
   use a distributional critic so it's not dmpo. It's also
   at the moment incompatible with the CRR agent where we use
   a double critic architecture. We can probably
   implement a generalized version of the mpo agent which allows
   users to choose between different critics.
3. The networks match what's being used in dmpo. Most importantly,
   by default, it's a large network using layernorm and mvn diag head
   that does not squash the mvn. Out of bounds actions is penalized
   in the mpo policy loss.

### CODE REVIEW

**Review by jacquelinekay — COMMENTED** [review]
The implementation looks right to me, I made some suggestions based on tricks I've seen for MPO before and will take a closer look after we discuss.

File: magi/agents/mpo/agent.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        networks,
+        config: mpo_config.MPOConfig,
+        seed: int,
+        device_prefetch: bool = True,
+        counter: Optional[counting.Counter] = None,
+        logger: Optional[loggers.Logger] = None,
+    ):
+        # This prevents reverb from deadlocking
+        min_replay_size = config.min_replay_size
+        # Local layout (actually agent.Agent) makes sure that we populate the
```
[the remark this answers is not in the document]
Comment (ethanluoyc) [reply magi/agents/mpo/agent.py:34]: Will fork LocalLayout from Acme and use that as the base class.

File: magi/agents/mpo/agent.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        config.min_replay_size = 1
+        config.min_observations = min_replay_size
+        self.builder = builder.MPOBuilder(config, logger_fn=lambda: logger)
+
+        random_key = jax.random.PRNGKey(seed)
+        learner_key, actor_key = jax.random.split(random_key)
+
+        # Setup reverb
+        replay_tables = self.builder.make_replay_tables(environment_spec)
+        replay_server = reverb.Server(replay_tables, port=None)
```
[the remark this answers is not in the document]
Comment (ethanluoyc) [reply magi/agents/mpo/agent.py:49]: yep

File: magi/agents/mpo/config.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    replay_table_name: str = adders_reverb.DEFAULT_PRIORITY_TABLE
+    prefetch_size: Optional[int] = None
+    samples_per_insert: float = 256
+
+    discount: float = 0.99
+    batch_size: int = 256
+    num_samples: int = 20
+    n_step: int = 5
+    clipping: bool = True
+    policy_learning_rate = 1e-4
```
[the remark this answers is not in the document]
Comment (ethanluoyc) [reply magi/agents/mpo/config.py:23]: I will leave these as the same as the default in Acme's MPO, but I do see these different from what's used in the paper.

File: magi/agents/mpo/learning.py [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+        policy_loss_fn: Optional[losses.MPO] = None,
+        counter: Optional[counting.Counter] = None,
+        logger: Optional[loggers.Logger] = None,
+    ):
+        policy_loss_fn: losses.MPO = policy_loss_fn or losses.MPO(
+            epsilon=1e-1,
+            epsilon_penalty=1e-3,
+            epsilon_mean=1e-3,
+            epsilon_stddev=1e-6,
+            init_log_temperature=1.0,
```
[the remark this answers is not in the document]
Comment (ethanluoyc) [reply magi/agents/mpo/learning.py:63]: Thanks for sharing this. I will keep this the same as what Acme has for now, if after running with a few examples I different settings work better, then I will update the defaults here accordingly.

File: magi/agents/mpo/learning.py [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+        discount: float,
+        num_samples: int,
+        action_dim: int,
+        target_policy_update_period: int,
+        target_critic_update_period: int,
+        policy_loss_fn: Optional[losses.MPO] = None,
+        counter: Optional[counting.Counter] = None,
+        logger: Optional[loggers.Logger] = None,
+    ):
+        policy_loss_fn: losses.MPO = policy_loss_fn or losses.MPO(
```
[the remark this answers is not in the document]
Comment (ethanluoyc) [reply magi/agents/mpo/learning.py:58]: Let me think about how to do this. Right now to support squashed action dist I will need to explicitly support a custom mpo loss, maybe I should add an additional bool flag to allow easy overwrite to use no action penalization.

File: magi/agents/mpo/learning.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ethanluoyc) [reply magi/agents/mpo/learning.py:63]: Adding a note to myself, 

There are some reference hyperparameters here https://arxiv.org/abs/2103.17258, which is different from the default used in tf.mpo

File: magi/agents/mpo/agent_test.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        spec = specs.make_environment_spec(environment)
+
+        # Make network purely functional
+        agent_networks = networks.make_networks(
+            spec,
+            policy_layer_sizes=(32, 32),
+            critic_layer_sizes=(32, 32),
+        )
+
+        # Construct the agent.
```
[the remark this answers is not in the document]
Comment (ethanluoyc) [reply magi/agents/mpo/agent_test.py:30]: Sounds good, will leave this as it is for now.

File: magi/agents/mpo/learning.py [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+                critic_params, o_tm1, transitions.action
+            )  # [B]
+            # q_tm1 = tf.squeeze(q_tm1, axis=-1)  # [B]; necessary for trfl.td_learning.
+
+            # Critic loss.
+            batch_td_learning = jax.vmap(rlax.td_learning)
+            td_error = batch_td_learning(
+                q_tm1, transitions.reward, discount * transitions.discount, q_t
+            )
+            critic_loss = jnp.mean(jnp.square(td_error))
```
[the remark this answers is not in the document]
Comment (ethanluoyc) [reply magi/agents/mpo/learning.py:116]: That's interesting, I have seen some sort of filtering when working with sequences but usually, those are done in the policy. 

Do you mean something like?
```
critic_loss = jnp.mean(jnp.square(td_error) * transitions.discount)
```
This would filter out the the transitions into a terminal state.
